include/aws/cal/symmetric_cipher.h

```diff
@@ -0,0 +1,263 @@
+    int (*finalize_decryption)(struct aws_symmetric_cipher *cipher, struct aws_byte_buf *out);
+};
+
+struct aws_symmetric_cipher {
+    struct aws_allocator *allocator;
+    struct aws_symmetric_cipher_vtable *vtable;
+    struct aws_byte_buf iv;
+    struct aws_byte_buf key;
+    struct aws_byte_buf aad;
+    struct aws_byte_buf tag;
+    size_t block_size;
+    size_t key_length_bits;
+    bool good;
+    void *impl;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/**
+ * Creates an instance of AES CBC with 256-bit key.
+ * If key and iv are NULL, they will be generated internally.
+ * You can get the generated key and iv back by calling:
+ *
+ * aws_symmetric_cipher_get_key() and
+ * aws_symmetric_cipher_get_initialization_vector()
+ *
+ * respectively.
+ *
+ * If they are set, that key and iv will be copied internally and used by the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_cbc_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv);
+
+/**
+ * Creates an instance of AES CTR with 256-bit key.
+ * If key and iv are NULL, they will be generated internally.
+ * You can get the generated key and iv back by calling:
+ *
+ * aws_symmetric_cipher_get_key() and
+ * aws_symmetric_cipher_get_initialization_vector()
+ *
+ * respectively.
+ *
+ * If they are set, that key and iv will be copied internally and used by the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_ctr_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv);
+
+/**
+ * Creates an instance of AES GCM with 256-bit key.
+ * If key, iv are NULL, they will be generated internally.
+ * You can get the generated key and iv back by calling:
+ *
+ * aws_symmetric_cipher_get_key() and
+ * aws_symmetric_cipher_get_initialization_vector()
+ *
+ * respectively.
+ *
+ * If they are set, that key and iv will be copied internally and used by the cipher.
+ *
+ * If tag and aad are set they will be copied internally and used by the cipher.
+ * decryption_tag would most likely be used for a decrypt operation to detect tampering or corruption.
+ * The Tag for the most recent encrypt operation will be available in:
+ *
+ * aws_symmetric_cipher_get_tag()
+ *
+ * If aad is set it will be copied and applied to the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_gcm_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv,
+    const struct aws_byte_cursor *aad,
+    const struct aws_byte_cursor *decryption_tag);
```

Comment:
This is a lot of arguments. Is there a good chance we'll add more in the future?
if so, consider an options-struct

Reply:
there's not a good chance on these crypto algorithms, and it's a "_new()" function and any version of it needing new arguments will be special enough to warrant another new function.